### PR TITLE
Optimise Distribution.Version.mkVersion

### PR DESCRIPTION
**Pull request by hvr**
NB: I'm submitting this PR on behalf of @dcoutts ; be careful when merging this PR to retain proper authorship

See https://github.com/dcoutts/cabal-version-bench/blob/master/src/Distribution/Version.hs

### CODE REVIEW

File: Cabal/Distribution/Version.hs [hunk 1]
```diff
@@ -169,24 +169,55 @@ instance Text Version where
...
+mkWord64VerRep :: Int -> Int -> Int -> Int -> Word64
+mkWord64VerRep v1 v2 v3 v4 =
+      (fromIntegral v1 `shiftL` 48)
+  .|. (fromIntegral v2 `shiftL` 32)
+  .|. (fromIntegral v3 `shiftL` 16)
+  .|.  fromIntegral v4
+
+{-# INLINE inWord16 #-}
+inWord16 :: Int -> Bool
+inWord16 x = (fromIntegral x :: Word) <= 0xffff
```
Comment (dcoutts) [inline Cabal/Distribution/Version.hs]: Need to import Word from Data.Word

File: Cabal/Distribution/Version.hs [hunk 2]
```diff
@@ -169,24 +169,55 @@ instance Text Version where
...
-    [v1,v2]       | v1 <= 0xfffe, v2 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) (v2+1) 0 0)
-    [v1,v2,v3]    | v1 <= 0xfffe, v2 <= 0xfffe, v3 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) (v2+1) (v3+1)  0)
-    [v1,v2,v3,v4] | v1 <= 0xfffe, v2 <= 0xfffe, v3 <= 0xfffe, v4 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) (v2+1) (v3+1) (v4+1))
-    v1:vs -> PV1 v1 vs
+mkVersion []                    = nullVersion
+mkVersion (v1:[])
+  | inWord16VerRep1 v1          = PV0 (mkWord64VerRep1 v1)
```
Comment (dcoutts) [inline Cabal/Distribution/Version.hs]: I don't understand. How did this pass the tests previously? Surely it should have gone wrong for negative v1? Are our tests ok?

Comment (hvr) [inline Cabal/Distribution/Version.hs]: Negative values are considered invalid input to `mkVersion` and hence undefined behaviour; consequently the tests made sure to test only with non-negative `[Int]`s...

File: Cabal/Distribution/Version.hs (hunk 1 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: bypassed via e14da30

File: Cabal/Distribution/Version.hs (hunk 2 above)
Comment (dcoutts) [inline Cabal/Distribution/Version.hs]: Ok

**Comment by dcoutts** [comment]
LGTM.  @23Skidoo merge if you're happy

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Pull request merged**